File: frontend/src/services/api/portal.ts
import type { Order } from '../../types/order';
import type { EligibleNode, SubscriptionItem } from '../../types/subscription';
import { API_BASE_URL, parseError } from './client';

export async function fetchMySubscriptions(token: string): Promise<SubscriptionItem[]> {
  const response = await fetch(`${API_BASE_URL}/api/v1/portal/subscriptions`, {
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    const err = await parseError(response, 'Failed to fetch customer subscriptions');
    throw new Error(err);
  }

  return response.json();
}

export async function fetchEligibleNodes(token: string, subId: number): Promise<EligibleNode[]> {
  const response = await fetch(`${API_BASE_URL}/api/v1/portal/subscriptions/${subId}/eligible-nodes`, {
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    const err = await parseError(response, 'Failed to fetch eligible nodes');
    throw new Error(err);
  }

  return response.json();
}

export async function switchSubscriptionNode(
  token: string,
  subId: number,
  targetNodeId: number
): Promise<SubscriptionItem> {
  const response = await fetch(`${API_BASE_URL}/api/v1/portal/subscriptions/${subId}/switch-node`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ target_node_id: targetNodeId }),
  });

  if (!response.ok) {
    const err = await parseError(response, 'Failed to switch server node');
    throw new Error(err);
  }

  return response.json();
}

export async function renewSubscription(
  token: string,
  subId: number,
  planId?: number
): Promise<Order> {
  const response = await fetch(`${API_BASE_URL}/api/v1/portal/subscriptions/${subId}/renew`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(planId ? { plan_id: planId } : {}),
  });

  if (!response.ok) {
    const err = await parseError(response, 'Failed to create renewal order');
    throw new Error(err);
  }

  return response.json();
}

File: backend/app/api/v1/portal.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.order import OrderResponse
from app.schemas.subscription import (
    EligibleNodeResponse,
    RenewSubscriptionRequest,
    SubscriptionResponse,
    SwitchNodeRequest,
)
from app.services.order_service import to_order_response
from app.services.portal_service import (
    create_renewal_order_for_subscription,
    get_customer_subscriptions,
    get_eligible_nodes_for_switch,
    switch_subscription_node,
)

router: APIRouter = APIRouter(
    prefix="/portal",
    tags=["customer-portal"],
    dependencies=[Depends(get_current_user)],
)


@router.get("/subscriptions", response_model=list[SubscriptionResponse])
async def list_my_subscriptions_endpoint(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[SubscriptionResponse]:
    """Retrieve all subscriptions owned by currently authenticated customer."""
    return await get_customer_subscriptions(db, user_id=current_user.id)


@router.get("/subscriptions/{sub_id}/eligible-nodes", response_model=list[EligibleNodeResponse])
async def list_eligible_nodes_endpoint(
    sub_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[EligibleNodeResponse]:
    """List available nodes with free capacity in same region for switching."""
    return await get_eligible_nodes_for_switch(db, subscription_id=sub_id, user_id=current_user.id)


@router.post("/subscriptions/{sub_id}/switch-node", response_model=SubscriptionResponse)
async def switch_node_endpoint(
    sub_id: int,
    payload: SwitchNodeRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SubscriptionResponse:
    """Migrate subscription to new node via gRPC while retaining client UUID & token."""
    return await switch_subscription_node(
        db,
        subscription_id=sub_id,
        user_id=current_user.id,
        target_node_id=payload.target_node_id,
    )


@router.post("/subscriptions/{sub_id}/renew", response_model=OrderResponse)
async def renew_subscription_endpoint(
    sub_id: int,
    payload: RenewSubscriptionRequest | None = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> OrderResponse:
    """Create renewal Order for customer subscription with VietQR payment instructions."""
    plan_id = payload.plan_id if payload else None
    order = await create_renewal_order_for_subscription(
        db,
        subscription_id=sub_id,
        user_id=current_user.id,
        plan_id=plan_id,
    )
    return await to_order_response(order, db)
